### PR TITLE
Add API to generate Stream from memory Address

### PR DESCRIPTION
**Pull request by rnjtranjan**
fixes:https://github.com/composewell/streamly/issues/1729

### CODE REVIEW

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 1]
```diff
@@ -140,3 +152,26 @@ fromFoldable = fromStreamK . K.fromFoldable
 {-# INLINE fromFoldableM #-}
 fromFoldableM :: (Monad m, Foldable f) => f (m a) -> Stream m a
 fromFoldableM = Prelude.foldr Stream.consM Stream.nil
+
+{-# INLINE ptrtoStreamK #-}
+ptrtoStreamK :: forall m a. (MonadIO m, Storable a) =>
+  Ptr a -> Int -> K.Stream m a
+ptrtoStreamK = go
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:159]: This should be in `StreamK` and should be called `fromPtrN` or `fromPtrNUnsafe`

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 2]
```diff
@@ -140,3 +152,26 @@ fromFoldable = fromStreamK . K.fromFoldable
...
+      case count of
+        0 -> K.nil
+        _ ->
+          let elemM = liftIO $ peek p
+            in elemM `K.consM` go (PTR_NEXT(p, a)) (count - 1)
+
+fromAddrN# :: forall m a. (MonadIO m, Storable a) =>
+  Int -> Addr# -> Stream m a
+fromAddrN# n addr =
+    Stream.fromStreamK $ ptrtoStreamK (castPtr $ Ptr addr) n
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:173]: `$` inside `(` `)` looks weird. Maybe,
```suggestion
    Stream.fromStreamK $ ptrtoStreamK (castPtr (Ptr addr)) n
```

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 3]
```diff
@@ -140,3 +152,26 @@ fromFoldable = fromStreamK . K.fromFoldable
...
+  where
+
+  go p count =
+      case count of
+        0 -> K.nil
+        _ ->
+          let elemM = liftIO $ peek p
+            in elemM `K.consM` go (PTR_NEXT(p, a)) (count - 1)
+
+fromAddrN# :: forall m a. (MonadIO m, Storable a) =>
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:170]: Add `Inline` statement

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 4]
```diff
@@ -27,13 +27,20 @@ module Streamly.Internal.Data.Stream.Generate
...
+
+    -- * From a memory address
+    , fromAddrN#
+    , fromPtrN
     )
 where
 
+import Control.Monad.IO.Class (MonadIO)
+import GHC.Exts (Addr#, Ptr (Ptr))
+import Foreign.Storable (Storable)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:39]: Sort order.

File: core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs [hunk 5]
```diff
@@ -459,3 +465,15 @@ fromListM = Stream step
     step _ (m:ms) = m >>= \x -> return $ Yield x ms
     step _ []     = return Stop
 #endif
+
+{-# INLINE fromPtrN #-}
+fromPtrN :: forall m a. (MonadIO m, Storable a) =>
+  Ptr a -> Int -> Stream m a
+fromPtrN p0 n0 = Stream step (p0, n0)
+
+  where
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs:474]: indent by 4 spaces and add a blank line after this.

File: core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs [hunk 6]
```diff
@@ -101,6 +105,8 @@ import qualified Streamly.Internal.Data.Unfold.Enumeration as Unfold
 
 import Prelude hiding (iterate, repeat, replicate, takeWhile)
 import Streamly.Internal.Data.Stream.StreamD.Type
+import Foreign.Storable (Storable (peek), sizeOf)
+import Foreign.Ptr (Ptr, plusPtr)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs:109]: sort order.

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 7]
```diff
@@ -140,3 +147,25 @@ fromFoldable = fromStreamK . K.fromFoldable
...
+-- XXX Should we rename this to fromAddrNUnsafe?
+{-# INLINE fromAddrN# #-}
+fromAddrN# :: (MonadIO m, Storable a) => Int -> Addr# -> Stream m a
+fromAddrN# n addr =
+    -- XXX You probably don't need a castPtr here
+    -- XXX Should look something like the below
+    -- Stream.fromStreamK $ StreamK.fromPtrNUnsafe (Ptr addr) n
+    Stream.fromStreamK $ ptrtoStreamK (Ptr addr) n
+
+-- XXX Should we rename this to fromPtrNUnsafe?
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:167]: When you are dealing with Ptrs you know what you are doing. We can add a warning in the docs.

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 8]
```diff
@@ -140,3 +147,25 @@ fromFoldable = fromStreamK . K.fromFoldable
 {-# INLINE fromFoldableM #-}
 fromFoldableM :: (Monad m, Foldable f) => f (m a) -> Stream m a
 fromFoldableM = Prelude.foldr Stream.consM Stream.nil
+
+-- XXX This should be in the StreamK module and should be named fromPtrN or
+-- fromPtrNUnsafe
+{-# INLINE ptrtoStreamK #-}
+ptrtoStreamK :: forall m a. (MonadIO m, Storable a) =>
+  Ptr a -> Int -> K.Stream m a
+ptrtoStreamK p = D.toStreamK . D.fromPtrN p
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:156]: Remove this.

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 9]
```diff
@@ -140,3 +147,25 @@ fromFoldable = fromStreamK . K.fromFoldable
...
+    -- XXX You probably don't need a castPtr here
+    -- XXX Should look something like the below
+    -- Stream.fromStreamK $ StreamK.fromPtrNUnsafe (Ptr addr) n
+    Stream.fromStreamK $ ptrtoStreamK (Ptr addr) n
+
+-- XXX Should we rename this to fromPtrNUnsafe?
+{-# INLINE fromPtrN #-}
+fromPtrN :: (MonadIO m, Storable a) => Int -> Ptr a -> Stream m a
+-- XXX fromPtrN n ptr = Stream.fromStreamK $ fromPtrNUnsafe ptr n
+fromPtrN n ptr = Stream.fromStreamK $ ptrtoStreamK ptr n
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:171]: This should be simply `Stream.fromStreamD . D.fromPtrN n`

File: core/src/Streamly/Internal/Data/Stream/Generate.hs [hunk 10]
```diff
@@ -140,3 +147,25 @@ fromFoldable = fromStreamK . K.fromFoldable
...
+ptrtoStreamK p = D.toStreamK . D.fromPtrN p
+
+-- XXX Should we rename this to fromAddrNUnsafe?
+{-# INLINE fromAddrN# #-}
+fromAddrN# :: (MonadIO m, Storable a) => Int -> Addr# -> Stream m a
+fromAddrN# n addr =
+    -- XXX You probably don't need a castPtr here
+    -- XXX Should look something like the below
+    -- Stream.fromStreamK $ StreamK.fromPtrNUnsafe (Ptr addr) n
+    Stream.fromStreamK $ ptrtoStreamK (Ptr addr) n
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/Generate.hs:165]: This should be simply `fromPtrN n (Ptr addr)`. Define this after fromPtrN.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs [hunk 11]
```diff
@@ -459,3 +465,18 @@ fromListM = Stream step
     step _ (m:ms) = m >>= \x -> return $ Yield x ms
     step _ []     = return Stop
 #endif
+
+-- | Convert a pointer to a 'Stream'. The actual number of
+-- elements reachable thru pointer is not checked.
+{-# INLINE fromPtrN #-}
+fromPtrN :: forall m a. (MonadIO m, Storable a) =>
+    Int -> Ptr a -> Stream m a
+fromPtrN n0 p0 = Stream step (p0, n0)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Stream/StreamD/Generate.hs:474]: Let's implement this as an unfold as well, in the Unfold module.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Review by rnjtranjan — COMMENTED** [review]
looks good.

**Pull request merged**